light-client-detector/src/examine.rs

```diff
@@ -0,0 +1,163 @@
+    trace: &Trace,
+    target_block: &LightBlock,
+    source: &Provider,
+) -> Result<(Trace, LightBlock), Error> {
+    let trusted_block = trace.first();
+
+    if target_block.height() < trusted_block.height() {
+        return Err(Error::target_block_lower_than_trusted(
+            target_block.height(),
+            trusted_block.height(),
+        ));
+    }
+
+    let mut previously_verified_block: Option<LightBlock> = None;
+
+    for trace_block in trace.iter() {
+        let result = examine_conflicting_header_against_trace_block(
+            source,
+            trace_block,
+            target_block,
+            previously_verified_block,
+        )?;
+
+        match result {
+            ExaminationResult::Continue(prev_verified_block) => {
+                previously_verified_block = Some(prev_verified_block);
+                continue;
+            },
+            ExaminationResult::Divergence(source_trace, trace_block) => {
+                return Ok((source_trace, trace_block));
+            },
+        }
+    }
+
+    // We have reached the end of the trace. This should never happen. This can only happen if one of the stated
+    // prerequisites to this function were not met.
+    // Namely that either trace[len(trace)-1].Height < targetBlock.Height
+    // or that trace[i].Hash() != targetBlock.Hash()
+    Err(Error::no_divergence())
+}
+
+#[derive(Debug)]
+pub enum ExaminationResult {
+    Continue(LightBlock),
+    Divergence(Trace, LightBlock),
+}
+
+fn examine_conflicting_header_against_trace_block(
+    source: &Provider,
+    trace_block: &LightBlock,
+    target_block: &LightBlock,
+    prev_verified_block: Option<LightBlock>,
+) -> Result<ExaminationResult, Error> {
+    // This case only happens in a forward lunatic attack. We treat the block with the
+    // height directly after the targetBlock as the divergent block
+    if trace_block.height() > target_block.height() {
+        // sanity check that the time of the traceBlock is indeed less than that of the targetBlock. If the trace
+        // was correctly verified we should expect monotonically increasing time. This means that if the block at
+        // the end of the trace has a lesser time than the target block then all blocks in the trace should have a
+        // lesser time
+        if trace_block.time() > target_block.time() {
+            return Err(Error::trace_block_after_target_block(
+                trace_block.time(),
+                target_block.time(),
+            ));
+        }
+
+        // Before sending back the divergent block and trace we need to ensure we have verified
+        // the final gap between the previouslyVerifiedBlock and the targetBlock
+        if let Some(prev_verified_block) = &prev_verified_block {
+            if prev_verified_block.height() != target_block.height() {
+                let source_trace =
+                    verify_skipping(source, prev_verified_block.clone(), target_block.clone())?;
```

Comment:
There is a call ~5 levels deep to the `verify()` function that calls `validate_against_trusted()`. I am not sure this needs to be called when coming from this calling context (i.e. misbehaviour).

Reply:
I think calling `verify()` here should be fine as the `now` gets propagated from the timestamp of latest block on the same chain. The problem occurs only when verifying misbehaviour headers in the IBC on-chain handler and that should be addressed by #1294